src/api/cvc4cpp.h

```diff
@@ -0,0 +1,2357 @@
+     * @return true if the sort is a datatype sort
+     */
+    bool isDatatype() const;
+
+    /**
+     * Is this a function sort?
+     * @return true if the sort is a function sort
+     */
+    bool isFunction() const;
+
+    /**
+     * Is this a predicate sort?
+     * That is, is this a function sort mapping to Boolean? All predicate
+     * sorts are also function sorts.
+     * @return true if the sort is a predicate sort
+     */
+    bool isPredicate() const;
+
+    /**
+     * Is this a tuple sort?
+     * @return true if the sort is a tuple sort
+     */
+    bool isTuple() const;
+
+    /**
+     * Is this a record sort?
+     * @return true if the sort is a record sort
+     */
+    bool isRecord() const;
+
+    /**
+     * Is this an array sort?
+     * @return true if the sort is a array sort
+     */
+    bool isArray() const;
+
+    /**
+     * Is this a Set sort?
+     * @return true if the sort is a Set sort
+     */
+    bool isSet() const;
+
+    /**
+     * Is this a sort kind?
+     * @return true if this is a sort kind
+     */
+    bool isSort() const;
+
+    /**
+     * Is this a sort constructor kind?
+     * @return true if this is a sort constructor kind
+     */
+    bool isSortConstructor() const;
+
+    /**
+     * @return the underlying datatype of a datatype sort
+     */
+    Datatype getDatatype() const;
+
+    /**
+     * Instantiate a parameterized datatype/sort sort.
+     * Create sorts parameter with Solver::mkParamSort().
+     * @param params the list of sort parameters to instantiate with
+     */
+    Sort instantiate(const std::vector<Sort>& params) const;
+
+    /**
+     * Output a string representation of this sort to a given stream.
+     * @param out the output stream
+     */
+    void toStream(std::ostream& out) const;
+
+    /**
+     * @return a string representation of this sort
+     */
+    std::string toString() const;
+
+  private:
+    /* The interal type wrapped by this sort. */
+    CVC4::Type* d_type;
+};
+
+/**
+ * Serialize a sort to given stream.
+ * @param out the output stream
+ * @param s the sort to be serialized to the given output stream
+ * @return the output stream
+ */
+std::ostream& operator<< (std::ostream& out, const Sort& s) CVC4_PUBLIC;
+
+/**
+ * Hash function for Sorts.
+ */
+struct CVC4_PUBLIC SortHashFunction
+{
+  size_t operator()(const Sort& s) const;
+};
+
+
+/* -------------------------------------------------------------------------- */
+/* Term                                                                       */
+/* -------------------------------------------------------------------------- */
+
+/**
+ * A CVC4 Term.
+ */
+class CVC4_PUBLIC Term
+{
+  friend class Solver;
+  friend struct TermHashFunction;
+
+  public:
+    /**
+     * Constructor.
+     */
+    Term();
+
+    /**
+     * Constructor.
+     * @param e the internal expression that is to be wrapped by this term
+     * @return the Term
+     */
+    Term(const CVC4::Expr& e);
```

Comment:
Why go to Expr instead of Node? (Skipping the Expr layer would handle the performance implications of Term.)

Reply:
Because SmtEngine is on the Expr layer. All expression arguments to functions of SmtEngine are of type Expr.

Skipping the Expr layer in this PR would mean that every Node argument that needs to be passed to SmtEngine has to be wrapped into a new Expr and every Expr returned from SmtEngine has to be unwrapped. @barrettcw what do you think? Should we skip the Expr layer?

Refactoring SmtEngine to use Nodes is out of scope for an initial version of the API imo (and would also break the current API).

Reply:
That is a fairly good argument. It makes sense to keep this clean in v1.

Reply:
Yes I agree - we can look into eliminating Expr later, but for now it makes sense to use Expr.